Adding view full log in alerts button

nimrodshn opened the pull request:
Currently we truncate (after 200 characters) the log error / warning that was received when connecting to a provider - In this patch I purpose a way of dealing with the **full** log error by adding a button to the alert when the alert is too long (above 200 characters).
Summary of scenarios tested:
Short alert:
![short_alert](https://user-images.githubusercontent.com/8366181/34154169-3e223c6e-e4bd-11e7-9c42-2113ed1939fc.gif)
Long alert:
![middle_alert](https://user-images.githubusercontent.com/8366181/34154181-485c86e4-e4bd-11e7-8d0c-7107ea619c8d.gif)
Very Long alert
![long_alert](https://user-images.githubusercontent.com/8366181/34154199-58fc6fdc-e4bd-11e7-8d4b-318ef66a1d3a.gif)


BZ: https://bugzilla.redhat.com/show_bug.cgi?id=1506718
cc: @cben , @serenamarie125 , @himdel 
Please share thoughts on this 😅  - this is just a proposition 😇

nimrodshn commented:
@miq-bot add_label bug, gaprindashvili/yes, wip

nimrodshn commented:
Depending on https://github.com/ManageIQ/manageiq/pull/16550

himdel commented:
This needs to be different.

You are changing the styling of every single flash message generated on JS side, in a way that is inconsistent with flash messages coming from ruby.

You're also randomly hardcoding a 200 character limit - this is wrong because this will affect any existing messages long enough. And it is also wrong because the reasonable size will depend on the user's resolution, and chosen language (this is why Twitter still limits Chinese to 140 chars).

Also, the CSS looks like it can't work on mobile, limiting height is seldom a good idea.

---

So.. proposing: 

* keep the both implementations consistent, and update ruby-side `add_flash` and corresponding templates to keep generating the same layout

* do a proper css class for the style changes, and ... if you can prevent using `overflow:hidden` and setting height, even better. (Test this on a phone.)

* don't determine the Details logic by message length, use an explicit option (feel free to add a 3rd options={} arg to add_flash if needed), apply only to messages coming from authentication mixin

(*or* we could add a separate message body concept to flash messages - so that we keep the flash the same, but can optionally add a longer body, which will trigger a button to display it, and needs no extra logic)

serenamarie125 commented:
@nimrodshn Although we don't have PatternFly support for this, I'd suggest doing something like OpenShift has implement ... where there is a See All link which allows for expansion and a vertical scrolllbar
![image](https://user-images.githubusercontent.com/12504403/33381628-8e0e7c1e-d4ec-11e7-9a71-1b287efb1c4a.png)

Tagging @jeff-phillips-18 who may be able to point you to some code which can does this

nimrodshn commented:
@himdel is @serenamarie125 's approach OK with you? (before I continue with implementation I just wanna make sure we're all on the same page).

himdel commented:
Depends on the details - @serenamarie125 's screenshot is a toast notification, not a flash message.

So... if you do manage to make the error go via our asynchronous notifications so that it ends up in the notification area... even better, I do believe there is some support for expandable messages already.

But it means going through a completely different code path than a flash message, and the message pretty much needs to come from the server, we have no way of creating these toasts in JS now.

(Or rather, no clean way - if this is something that doesn't have to be visible in the notification area after the user reloads the page, then maybe we do. If it needs to be there after reload, it needs to come from the server.)

(I don't really object to the flash message approach either, just that the style needs to be consistent, and work for all the existing ones too.)

himdel commented:
@serenamarie125 Yes, it's the inline one - the one you can see in the gif in the description (except the button is new there).

serenamarie125 commented:
@nimrodshn I'd suggest using similar behavior, regardless if it's a flash/inline or toast notification.  

@himdel I think we need to discuss the flash/inline versus toast approach in general and see how we can transition away from flash/inline in most cases.  I wouldn't expect that to be part of this PR.

nimrodshn commented:
So @jrafanie defers for a different solution in the backend:

> I'd be ok with a few different solutions:
> - extending the truncate length to fit a reasonable error message to fix your bug (while still fitting into the status_details column)
> - Or move the truncate to the log line and where we save the status_details and all callers of authentication_check

I think outputting the log file is a good solution as it doesn't effect our current UI - Is that OK with you @himdel @serenamarie125 ?

himdel commented:
Re-reading the BZ, I'm wondering if we're looking at the problem the right way.. does the user really need to get that long HTML message in its entirety? Do we assume the user will be able to understand it?

Because, if this is not about getting randomly long descriptive messages from backend, but about sometimes getting unexpected HTML back, maybe we could fix this by actually dealing with the HTML on the backend - parsing it, or replacing it with a simpler message.

(I don't know the details of what openshift actually returns when, but seems to me that if we just stripped all the html tags and left just the text nodes, if would probably fit in the limit while still letting the user know.)

(Incidentally, that's what we do in JS when the rails backend returns an unexpected response - try to parse out the exception from the message.)

((But if parsing it server-side doesn't work .. I don't know, my complaints were more about the style changes to the flash messages than anything else - if we drop the length limit without changing the format, I don't think there's any issue, as long as the backend is happy. :) ))

cben commented:
There are a couple common mistakes (such as "no such app" page from
openshift router).
But it's impossible to say what we'll get generally, we can't even know
whether it's HTML!  (for working openshift it's actually JSON)
The main scenario I had in mind when I opened that BZ is *talking to
something unexpected* — because user entered wrong hostname/port, or
entered correct ones but reached something else.
(In a customer case that prompted this, a firewall blocked the connection.)

The user might or might not understand it...  In the firewall case, first
200 chars was unhelpful but once I asked them to run `curl` and they saw
"blocked" and the firewall vendor name they went "aha!" and fixed it
quickly.

The way I see it, when you're trying to connect a provider and it Just
Doesn't Work (tm), it's a miserable experience.
At that point, you got exactly zero value of the product, you can't make
any progress until you get help, and you don't even have enough data for
troubleshooting, like text to google — or to send to people who might help.
(well in case of RedHat support you'd be asked to upload logs, but in any
case triage is faster if you have error message)

I like the idea of backend sending independent "message" and "details"
texts, because in future we might want to add more diagnostic data there,
e.g. details on the TLS certificate, maybe some advice / pointers to docs
etc...

nimrodshn commented:
@himdel PTAL

nimrodshn commented:
@himdel code climate is really annoying 😭

himdel commented:
@nimrodshn Feel free to ignore :)

We do need to fix CC at some point, yes.

---

The code looks good I think :+1:

(have not tested in UI yet)

One issue I see - we can't use an `id` for `alert_expand_button` .. you can have more than one flash message on the screen, so it needs to be a class.

Other than that, can you please also add the corresponding change to the Ruby `add_flash`?
(Or I can do a separate PR if you have trouble there, just want to ensure both are in sync.)

nimrodshn commented:
@himdel Sure will do that!

cben commented on app/controllers/mixins/ems_common_angular.rb:
```diff
@@ -98,7 +98,7 @@ def render_validation_result(result, details)
       end
 
-      render_flash_json(msg, level)
+      render_flash_json(msg, level, :long_alert => true)
```
Nowsg is unlimited, perhaps ruby should still truncate to some large limit here, to avoid sending megabyte responses to UI.

himdel replied:
Is this addressed by https://github.com/ManageIQ/manageiq/pull/16550/files#diff-f9cace1fe4fdcec66f962b23f2110fd6R337 now?

cben commented:
more like 20000 if you ask me :)
it's will be loooong but only in rare cases, and it's hidden by default,
and in those cases long is better than lack of info...
it's just to avoid completely blowing up...

(well a gigabytes response could still exhaust ruby memory before we even
get to UI.  but that's out of scope, let's hope people don't connect MIQ
directly to youtube :-)

nimrodshn commented:
@himdel I have added another further truncate of 20000 chars (as suggested by @cben) in the backend.
Also, do you mind doing another PR for the ruby `add_flash`? I'm having trouble with dynamic classes in `/layouts/_flash_msg.html.haml` plus I dont want this PR to get too big.

ohadlevy commented on app/assets/javascripts/miq_flash.js:
```diff
@@ -35,6 +39,15 @@ function add_flash(msg, level, options) {
   alertDiv.append(iconSpan, textStrong);
 
+  if (options.long_alert) {
+    var detailsButton = $('<div style="padding-top: 4px"><button id="alert_expand_button" class="btn btn-default pull-right" type="submit">Details</button></div>');
```
String should be extracted, also afaiu @serenamarie125 suggested to use the "see more" text vs "details"?

ohadlevy commented on app/assets/javascripts/miq_flash.js:
```diff
@@ -52,6 +65,16 @@ function clearFlash() {
 }
 
+function expandAlert(params) {
+    if (!params.clicked) {
```
inconsistent whitespace?

nimrodshn commented:
@ohadlevy thank you for you're review!
I have changed the button name to 'See More' (see screenshot)
Plus I have added tests and refactored accordingly.


Screenshot:
![screenshot from 2017-12-11 12-41-39](https://user-images.githubusercontent.com/8366181/33827388-de8fdda8-de70-11e7-8175-5af911237621.png)

nimrodshn commented:
@himdel PTAL.

nimrodshn commented:
@himdel I have added tests as requsted by @ohadlevy plus i have added the option of checking that the text is *not* overflowing even though the `long_response: true` is given (In which case the added button simply doesn't do anything.) so the button is disabled.

ohadlevy commented on app/assets/javascripts/miq_flash.js:
```diff
@@ -35,6 +39,17 @@ function add_flash(msg, level, options) {
 
+  // if options.long alert is given than add a `See More` button to the alert div to allow user the get more details of the error/alert.
+  if (options.long_alert) {
+    var detailsButtonTxt = "See More";
```
please extract text....
also it might not be in a bottom (/cc @serenamarie125 )

nimrodshn replied:
*button

ohadlevy commented on app/assets/javascripts/miq_flash.js:
```diff
@@ -46,12 +61,27 @@ function add_flash(msg, level, options) {
   $('#flash_msg_div').append(textDiv).show();
+
+  // Check if text is *not* overflowing and options.long_alert is true thus disabling the button.
+  if ((alertDiv[0].scrollWidth <= alertDiv.innerWidth())  && options.long_alert) {
```
nitpick: extra whitespace before ```&&```

nimrodshn commented:
@himdel @serenamarie125 There is also the option of having a link like so:
![screenshot from 2017-12-11 18-53-22](https://user-images.githubusercontent.com/8366181/33842944-b47a50fa-dea4-11e7-8ac5-95aec5b580e1.png)

As suggested by @yaacov .

nimrodshn commented:
@serenamarie125 Do you mind giving some more feedback on this effort? is it OK with you as it stands now?

himdel commented:
Removing `pending core` - https://github.com/ManageIQ/manageiq/pull/16550#issuecomment-350027983

himdel commented on app/assets/javascripts/miq_flash.js:
```diff
@@ -35,6 +39,17 @@ function add_flash(msg, level, options) {
+  // if options.long alert is given than add a `See More` button to the alert div to allow user the get more details of the error/alert.
+  if (options.long_alert) {
+    var detailsButtonTxt = _("See More");
+    var detailsButton = $('<div style="padding-top: 4px"><button id="alert_expand_button" class="btn btn-default pull-right" type="submit">' + detailsButtonTxt + '</button></div>');
```
We can't use `button` or `input type=submit`, because most often, a flash message is *inside* a form. So, instead of expanding, you'd expand *and* submit the form.

(Or, we need to explicitly `event.stopPropagation()`)

nimrodshn replied:
@himdel just the be clear - we cannot use button **at all**? or is it just the `type=submit` that's causing the issue here?

himdel replied:
Aah.. ok... `button` with `type=submit` is just a `button`, `type` is not actually a valid attribute on a button element, only on an input.

So... it's not that we can't use either, it's just that we'd have to stop the event from propagating any further.

But... most likely the best bet would be to use `<a href="" class="btn btn-default">` since that one will never submit a form.

EDIT: or btn-link if this is supposed to be a link now :)

nimrodshn replied:
@himdel nice! :+1:

himdel commented on app/assets/javascripts/miq_flash.js:
```diff
@@ -46,12 +61,27 @@ function add_flash(msg, level, options) {
   $('#flash_msg_div').append(textDiv).show();
+
+  // Check if text is *not* overflowing and options.long_alert is true thus disabling the button.
+  if ((alertDiv[0].scrollWidth <= alertDiv.innerWidth()) && options.long_alert) {
```
Is this bit strictly necessary?

Because, if we do that, we need to watch for `resize` events and recompute...

nimrodshn replied:
@himdel we can remove this altogether if we don't mind having a dangling button...  There is a case when we pass `long_alert => true` but the actual text is not large than we would have a button that does nothing...
If we wish to solve this case - I can try and find a different solution if you're not comfortable with this one.

himdel replied:
I'm just worried that if we do it, we'll get BZs that it won't work when the window gets resized :).

But you're right that if we just let the button active, we may equally well get BZs that say the button does nothing :).

Unless somebody has a strong opinion here, maybe just leave it as is and we'll see what happens.

himdel commented on app/assets/javascripts/miq_flash.js:
```diff
@@ -46,12 +61,27 @@ function add_flash(msg, level, options) {
+    $(".alert").css({height: "auto", 'text-overflow': "initial", "white-space": "normal"});
+    params.clicked = true;
+  } else {
+    $(".alert").css({overflow: "hidden", 'text-overflow': "ellipsis", "white-space": "nowrap"});
```
sounds like this is just..

```js
$('.alert').addClass('text-overflow-pf');
```

isn't it? And removeClass for the other case..

himdel replied:
Also.. you may need to pass in the specific element .. otherwise this will affect all the flash messages, not just the one..

serenamarie125 commented:
@nimrodshn you show 2 screenshots.  UX recommendation would be to show a link, not a button.  The link label should be "View More".  HTH

serenamarie125 commented:
@nimrodshn could you also include a screenshot of what it looks like when the user clicks "View More"?

nimrodshn commented:
@serenamarie125 I have just updated the GIF in the description ⬆️ .
@himdel PTAL.

ohadlevy commented:
@nimrodshn shouldn't the view more text change to view less after its expanded?
I also feel that the original length might be a bit too long already?

nimrodshn commented:
@ohadlevy will change to 'View Less`.
As for the other comment - its just one line..

ohadlevy commented:
how would that line look on a smaller display?

nimrodshn commented:
@ohadlevy it's still one line 😄 
![screenshot from 2017-12-14 12-02-05](https://user-images.githubusercontent.com/8366181/33986659-ec43f3aa-e0c6-11e7-9669-804c0d2f7362.png)

yaacov commented on app/assets/javascripts/miq_flash.js:
```diff
@@ -27,6 +27,10 @@ function add_flash(msg, level, options) {
       throw("add_flash: unknown level: " + level);
   }
 
+  if (options.long_alert) {
```
you should not depend on options here,
we can deduce that we need the new feature if  msg is longer then some value:
```
var SHORT_MSG_LENGTH = 120;
var shortMsg;

if (msg.length > SHORT_MSG_LENGTH) { ...
shortMsg = msg.substring(0, SHORT_MSG_LENGTH) + '...';
```

nimrodshn replied:
thing is @himdel didn't want to hardcode the length thats why we added the options...

yaacov replied:
It is still hard coded to `200` in [manageiq L94](https://github.com/ManageIQ/manageiq/pull/16550/files#diff-5915b4f8bc244828cd9b675fe145ee74R95) 

or am I missing something ?

nimrodshn replied:
See the comments above...
> You're also randomly hardcoding a 200 character limit - this is wrong because this will affect any existing messages long enough. And it is also wrong because the reasonable size will depend on the user's resolution, and chosen language (this is why Twitter still limits Chinese to 140 chars).

yaacov replied:
Currently the text in the line is too long ... so something is wrong here :-)
I would expect a flash to look like that:

Short view (only one or two lines in small screens)
```
Lorem ipsum dolor sit amet, per ne erat vitae sententiae, eam tollit sententiae ut, 
dicant eripuit suavitate an cum. Ea al...  Show More
```

Long view (show full message)
```
Lorem ipsum dolor sit amet, per ne erat vitae sententiae, eam tollit sententiae ut, 
dicant eripuit suavitate an cum. Ea alii elitr graecis has. Scripta iracundia cum eu, 
adipisci evertitur ea est. Mei in partem adversarium. Facete adipiscing necessitat
ibus qui at, an accusamus repudiare scripserit mea, odio dissentiunt ea sed. 

Show Less
```

@nimrodshn  @ohadlevy @himdel WDYT ?

nimrodshn replied:
> Short view (only one or two lines in small screens)
@yaacov it still only show's one line, how short do you want the msg to be? half a line?

yaacov replied:
@nimrodshn 
If `text-overflow: ellipsis;` does not work here (see gif in description) then you can go back to `trancate` by some constant, IMHO it's ok for this length to be more then one line on smaller screens, and it's also ok for this short version to be more short for some locales ( e.g. Chinese ), see [above example](https://github.com/ManageIQ/manageiq-ui-classic/pull/2862#discussion_r156909937).
@himdel WDYT ?

yaacov replied:
> @yaacov it still only show's one line (regardless of the screen) , how short do you want the msg to be? half a line?

see [above example](https://github.com/ManageIQ/manageiq-ui-classic/pull/2862#discussion_r156909937).
If you look carefully you can see I show two lines in that example.

yaacov replied:
> @yaacov it still only show's one line (regardless of the screen)

This is problematic on small devices, you sometimes want to show more then one line:

Big screen, one line is enough for short message: 
```
This is a problem that can be fixed easily by pressing this button
```

Small screen, you may want two or three lines for short message: 
```
This is a problem that
can be fixed easily by
pressing this button
```
If you show only one line:
```
This is a problem that...
```
This is not what I as user want to see :-)

himdel replied:
@yaacov Is this really a problem if the user can still expand the message to full length? Would your issue be solved if this was 2 lines instead of one?

As for https://github.com/ManageIQ/manageiq-ui-classic/pull/2862/files#r156909937 .. I'm not quite understanding what the issue is (or missing a screenshot where it goes wrong).

ohadlevy commented:
what about something like ellipsis instead?
https://css-tricks.com/snippets/css/truncate-string-with-ellipsis/

On Thu, Dec 14, 2017 at 12:46 PM, Yaacov Zamir <notifications@github.com>
wrote:

> *@yaacov* commented on this pull request.
> ------------------------------
>
> In app/assets/javascripts/miq_flash.js
> <https://github.com/ManageIQ/manageiq-ui-classic/pull/2862#discussion_r156909937>
> :
>
> > @@ -27,6 +27,10 @@ function add_flash(msg, level, options) {
>        throw("add_flash: unknown level: " + level);
>    }
>
> +  if (options.long_alert) {
>
> Currently the text in the line is too long ... so something is wrong here
> :-)
> I would expect a flash to look like that:
>
> Short view (only one or two lines in small screens)
>
> Lorem ipsum dolor sit amet, per ne erat vitae sententiae, eam tollit sententiae ut,
> dicant eripuit suavitate an cum. Ea al...  Show More
>
> Long view (show full message)
>
> Lorem ipsum dolor sit amet, per ne erat vitae sententiae, eam tollit sententiae ut,
> dicant eripuit suavitate an cum. Ea alii elitr graecis has. Scripta iracundia cum eu,
> adipisci evertitur ea est. Mei in partem adversarium. Facete adipiscing necessitat
> ibus qui at, an accusamus repudiare scripserit mea, odio dissentiunt ea sed.
>
> Show Less
>
> @nimrodshn <https://github.com/nimrodshn> @ohadlevy
> <https://github.com/ohadlevy> @himdel <https://github.com/himdel> WDYT ?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ManageIQ/manageiq-ui-classic/pull/2862#discussion_r156909937>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AABOx0GpJL8MenHNRm_6EPQVtFnwEfQLks5tAPxqgaJpZM4Qu7YH>
> .
>

nimrodshn commented:
@ohadlevy in this implementation  we are using `ellipsis` (thats under `text-overflow-pf`) - `ellipsis` truncates to one line like in here.

```
.text-overflow-pf {
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
  word-wrap: normal;
}
```

ohadlevy commented:
then why are we truncating server side as well?

On Thu, Dec 14, 2017 at 12:54 PM, Nimrod Shneor <notifications@github.com>
wrote:

> *@nimrodshn* commented on this pull request.
> ------------------------------
>
> In app/assets/javascripts/miq_flash.js
> <https://github.com/ManageIQ/manageiq-ui-classic/pull/2862#discussion_r156911764>
> :
>
> > @@ -27,6 +27,10 @@ function add_flash(msg, level, options) {
>        throw("add_flash: unknown level: " + level);
>    }
>
> +  if (options.long_alert) {
>
> Short view (only one or two lines in small screens)
> @yaacov <https://github.com/yaacov> it still only show's one line, how
> short do you want the msg to be? half a line?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ManageIQ/manageiq-ui-classic/pull/2862#discussion_r156911764>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AABOx2w3TZ2YxzntfxLFdA6t0EjzGTLjks5tAP5UgaJpZM4Qu7YH>
> .
>

yaacov commented:
> then why are we truncating server side as well?

@ohadlevy , @nimrodshn has a PR to remove this truncating, currently open but approved.

nimrodshn commented:
@ohadlevy The idea is that we allow long alert msg's to propagate from the backend and we try to deal with them elegantly in the UI.

nimrodshn commented:
@himdel What do you think of this approach:
```
var SHORT_MSG_LENGTH = 120;
var shortMsg;

if (msg.length > SHORT_MSG_LENGTH) { ...
shortMsg = msg.substring(0, SHORT_MSG_LENGTH) + '...';
```
Raised by @yaacov .

nimrodshn commented:
@ohadlevy updated the button to 'Show Less' when the alert is 'expanded'. (Updated the GIF ⬆️ ).

serenamarie125 commented:
@nimrodshn I'd suggest a vertical scrollbar be added to the inline notification when expanded if there are more than 5 total lines (including View Less link)

The View More / View Less links look good.

yaacov commented on app/assets/javascripts/miq_flash.js:
```diff
@@ -35,6 +39,17 @@ function add_flash(msg, level, options) {
 
+  // if options.long alert is given than add a `See More` button to the alert div to allow user the get more details of the error/alert.
+  if (options.long_alert) {
+    var detailsLinkTxt = _("View More");
```
do you add the "View More" even if msg is short ?

e.g.
```
short error msg. View More
```

nimrodshn replied:
@yaacov yep. We discussed it with @himdel before.. I agree it's not ideal.

yaacov replied:
Than why not check the length  ?

yaacov replied:
nice, @himdel agreed to show a do-nothing "View More"/"View Less" links, that only change from "More" to "Less" when pressed :-) , I will remember that one next time he nitpick me on something.

nimrodshn replied:
@yaacov I think I will check the length eventually.. I just want @himdel to agree to it too cause I know he has issues with this approach and he's the one that matters 😝

himdel replied:
This is now addressed by [checkEllipsis](https://github.com/ManageIQ/manageiq-ui-classic/pull/2862/files#diff-c31814e321dfd0b05558e69e59e39e68R75), right?

yaacov replied:
Yes :-)
links are not displayed when we do not need them 👍

himdel commented:
> @himdel What do you think of this approach:

@nimrodshn Definitely no hardcoded number of characters please.

That's worse not only when resizing, it's also worse for i18n :).

It was much better before, sorry :(

nimrodshn commented:
@himdel @yaacov @ohadlevy Changed the code to correspond to @yaacov & @ohadlevy 's comments - this also solved the dangling like situation that had occurred before when the line was short but `long_alert => true`:
![no-hardcodedline](https://user-images.githubusercontent.com/8366181/34109349-ed32805e-e40b-11e7-8902-15f74e51bda5.gif)
Still need to add scrollbar when msg is long.
@himdel PTAL

himdel commented:
@nimrodshn Sorry, if you can just put the "1 line" logic back instead of the character counting logic.

As for the scrollbar, should be relatively easy to limit max-height to a number of lines - you probably want `ex` as the CSS units, **not** pixels ;).

himdel commented:
Ok, there seems to have been some confusion introduced so..

Please note that from the start, I pushed @nimrodshn to keep this dependent on visual things like screen size, instead of arbitrary limits like a number of characters.

This was so that the text would always have a sane size limit, no matter the resolution, etc.

This worked, with the only thing not having been solved being what happens when the user manually resizes the screen after that. We chose to ignore it to see if that problem actually happens, or if somebody would have an idea on how to fix this better.

But going back to the original character counting implementation is a way in the wrong direction.

nimrodshn commented:
@himdel reverted to the old commit. just need to add the scrollbar.

nimrodshn commented:
@himdel Added the scroll bar plus fixed the dangling link.

yaacov commented on app/assets/javascripts/miq_flash.js:
```diff
@@ -46,12 +61,46 @@ function add_flash(msg, level, options) {
+    params.clicked = true;
+    params.alert_elem.removeClass('text-overflow-pf');
+    params.alert_elem.addClass('text-vertical-overflow-pf');
+    params.link.find('a').text("View Less");
```
translate ?

nimrodshn replied:
nice catch!

yaacov commented on app/assets/javascripts/miq_flash.js:
```diff
@@ -46,12 +61,46 @@ function add_flash(msg, level, options) {
+
+  // remove dangling 'Show More' link when the alert msg is short.
+  if( options.long_alert && !checkElipsis(alertDiv) ) {
+    detailsLink.hide();
```
nice 👍

nimrodshn commented:
cc @skateman  @serenamarie125 
GIF(with scrollbar): 
![long_alert](https://user-images.githubusercontent.com/8366181/34125985-c11f08fe-e440-11e7-9bb4-c5a877ac61e5.gif)

himdel commented on app/assets/javascripts/miq_flash.js:
```diff
@@ -46,12 +61,48 @@ function add_flash(msg, level, options) {
+}
+
+function expandAlert(params) {
+  var viewMoreTxt = _("View More");
```
In js, you need `__`, not `_` for i18n ;).

(`_` seems to work because it's lodash("string") -> ["string"].toString() -> "string" :))

serenamarie125 commented:
@miq-bot add_label ux/review

serenamarie125 commented:
@miq-bot add_label ux/approved

serenamarie125 commented:
@miq-bot remove_label ux/review

himdel commented:
Tested in UI, looks like resizing the window now works both for expanded and for non-expanded alerts :+1: 

Two more things I notice:

* if you have more than one alert, the code touches both whenever you click any of them
* checkEllipsis works in chrome but not in firefox (or at least not firefox 51) - View More is displayed there every time

himdel commented on app/assets/javascripts/miq_flash.js:
```diff
@@ -35,6 +39,17 @@ function add_flash(msg, level, options) {
+    var detailsLink = $('<div class="alert_expand_link"><strong><a href="#">' + detailsLinkTxt + '</a></strong></div>');
+    var params = {clicked: false, alert_elem: alertDiv, link: detailsLink};
+    $(document).on('click', '.alert_expand_link', function() {
+      expandAlert(params);
```
`expandAlert(params, this)`

himdel commented on app/assets/javascripts/miq_flash.js:
```diff
@@ -35,6 +39,17 @@ function add_flash(msg, level, options) {
+  if (options.long_alert) {
+    var detailsLinkTxt = _("View More");
+    var detailsLink = $('<div class="alert_expand_link"><strong><a href="#">' + detailsLinkTxt + '</a></strong></div>');
+    var params = {clicked: false, alert_elem: alertDiv, link: detailsLink};
```
You may want to save the state in that particular element's `data`, or somewhere per-element, not global.

himdel commented on app/assets/javascripts/miq_flash.js:
```diff
@@ -46,12 +61,48 @@ function add_flash(msg, level, options) {
+  var found = false;
+  var $c = element
+    .clone()
+    .css({display: 'inline', width: 'auto', visibility: 'hidden'})
```
`display: 'inline-block'` will make this work in firefox :)

simaishi commented:
Gaprindashvili backport details:

```
$ git log -1
commit 56283a047f9e3571cce1538abadd723a44b94f97
Author: Martin Hradil <himdel@seznam.cz>
Date:   Wed Dec 20 16:41:15 2017 +0100

    Merge pull request #2862 from nimrodshn/fix_log_bug
    
    Adding view full log in alerts button
    (cherry picked from commit dd0eadbacc34eab74dd1524e7492d8825bdaedbb)
    
    https://bugzilla.redhat.com/show_bug.cgi?id=1530645
```
